src/arith/solve_linear_equation.cc

```diff
@@ -0,0 +1,476 @@
+      index -= 1;
+    }
+  }
+}
+
+Map<Var, Range> InferRange(const Map<Var, PrimExpr>& vars_to_infer,
+                           const Array<Var>& ori_vars,
+                           const Map<Var, Range>& ori_ranges) {
+  // The resulting ranges
+  Map<Var, Range> new_ranges;
+
+  std::unordered_set<const VarNode*> ori_vset;
+  for (const Var& v : ori_vars) {
+    ori_vset.insert(v.get());
+  }
+
+  std::unordered_map<const VarNode*, IntSet> var_intsets;
+  for (const auto& p : ori_ranges) {
+    if (!ori_vset.count(p.first.get())) {
+      // First of all, fill the new ranges with outer variable ranges
+      new_ranges.Set(p.first, p.second);
+    }
+    // Convert original ranges to IntSets
+    var_intsets[p.first.get()] = IntSet::range(p.second);
+  }
+
+  // Infer ranges for the new variables and add them to the resulting ranges
+  for (const auto& p : vars_to_infer) {
+    const auto& var = p.first;
+    const auto& expr = p.second;
+    Range range = EvalSet(expr, var_intsets).cover_range(Range());
+    if (range.defined()) {
+      new_ranges.Set(var, range);
+    }
+  }
+  return new_ranges;
+}
+
+// pretty print matrix equation
+void DebugPrint(const std::vector<std::vector<int64_t>>& S,
+                const std::vector<std::vector<int64_t>>& V,
+                const std::vector<PrimExpr>& V_inv_x,
+                const std::vector<PrimExpr>& rhs) {
+  std::cout << "S:\n";
+  for (size_t i = 0; i < S.size(); ++i) {
+    for (auto e : S[i]) {
+      std::cout << e << "\t";
+    }
+    std::cout << "\t->\t" << rhs[i];
+    std::cout << "\n";
+  }
+  std::cout << "V:\n";
+  for (const auto& r : V) {
+    for (auto e : r) {
+      std::cout << e << "\t";
+    }
+    std::cout << "\n";
+  }
+  std::cout << "V_inv x:\n" << Array<PrimExpr>(V_inv_x);
+  std::cout << "\n" << std::endl;
+}
+
+LinearSystemTransform SolveEquations(const LinearSystem& system_to_solve) {
+  // m: # of equations
+  // n: # of variables
+  // we first construct A_{mxn} x_{nx1} = y_{mx1}
+  // then get Smith normal form of matrix A,
+  // S_{mxn} = U_{mxm} A_{mxn} V_{nxn}
+  // => U^{-1} S V^{-1} x = y
+  // S V^{-1} x = U y
+  std::vector<PrimExpr> Uy;  // mx1
+  std::vector<std::vector<int64_t>> S;  // mxn
+  std::vector<std::vector<int64_t>> V;  // nxn
+  std::vector<PrimExpr> V_inv_x;  // V^{-1} x, nx1
+  // Conditions we don't know what to do with
+  std::vector<PrimExpr> rest;
+
+  Analyzer analyzer_problem;
+  analyzer_problem.Bind(system_to_solve->ranges);
+
+  size_t num_vars = system_to_solve->variables.size();
+
+  // initialize V_{nxn} with identity matrix,
+  // initialize V^{-1} x as x
+  for (size_t i = 0; i < num_vars; ++i) {
+    V.emplace_back(num_vars);
+    V.back()[i] = 1;
+    V_inv_x.push_back(system_to_solve->variables[i]);
+  }
+
+  // Transform formulas into rows of the matrix
+  // S_{mxn} V^{-1}_{nxn} x_{nx1} = U y, in which n is # of variables
+  // here we initialize S_{mxn} to be A, U to be identity matrix.
```

Comment:
V to be identity matrix?

Reply:
V was described above. here what I mean is
U^{-1} S V^{-1} x = y = [coefs] => S V^{-1} x = U [coefs]
which implies U is identity.